### PR TITLE
[Cherry-Pick]: UefiCpuPkg/CpuMpPei. Use CpuPageTableLib to set memory attribute. Parallel get stack base for better performance.

### PR DESCRIPTION
**Pull request by apop5**
## Description

Cherry picking two changes
[Parallel get stack base for better performance.] https://github.com/tianocore/edk2/commit/e8166a852e63ee0c7c37291c1f88b19cc4692514)
[Use CpuPageTableLib to set memory attribute.](https://github.com/tianocore/edk2/commit/7e18c9a788e543ab71cdc0485989cf5d00cdccc2)

Cherry-picks originate for performance improvement recommended by Silicon Vendor. 

Added a Status check on new functionality to prevent codeql error. 

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [X] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested
Local CI runs
Test on local hardware

## Integration Instructions
N/A

### CODE REVIEW

**Comment by apop5** [comment]
@TaylorBeebe 

Can you please double check this cherry-pick?

I am worried that the settings they are using do not align with memory protections that will be applied later.

**Comment by TaylorBeebe** [comment]
> @TaylorBeebe
> 
> Can you please double check this cherry-pick?
> 
> I am worried that the settings they are using do not align with memory protections that will be applied later.

It looks like the end result is the same before and after this cherry-pick so it should be fine. DxeMpLib and MemoryProtectionSupport will take care of setting attributes on stack memory in DXE.

Is there something in particular I might be missing?

**Comment by apop5** [comment]
> > @TaylorBeebe
> > Can you please double check this cherry-pick?
> > I am worried that the settings they are using do not align with memory protections that will be applied later.
> 
> It looks like the end result is the same before and after this cherry-pick so it should be fine. DxeMpLib and MemoryProtectionSupport will take care of setting attributes on stack memory in DXE.
> 
> Is there something in particular I might be missing?

There is a allocate pages call now in Pei. As far as I can tell, its allocating boot services data, and I am worried that might throw off our expectations when we transition in the dxe phase.

**Comment by TaylorBeebe** [comment]
> > > @TaylorBeebe
> > > Can you please double check this cherry-pick?
> > > I am worried that the settings they are using do not align with memory protections that will be applied later.
> > 
> > 
> > It looks like the end result is the same before and after this cherry-pick so it should be fine. DxeMpLib and MemoryProtectionSupport will take care of setting attributes on stack memory in DXE.
> > Is there something in particular I might be missing?
> 
> There is a allocate pages call now in Pei. As far as I can tell, its allocating boot services data, and I am worried that might throw off our expectations when we transition in the dxe phase.

I think it should be fine. AllocatePageTableMemory() also redirects to using PEI services within MemoryAllocationLib so I think the final state is still the same.

**Review by TaylorBeebe — APPROVED** [review]

File: UefiCpuPkg/CpuMpPei/CpuPaging.c
Comment (makubacki) [inline UefiCpuPkg/CpuMpPei/CpuPaging.c:1]: It looks like you're trying to cherry-pick with an additional commit on top. You'll need to rebase locally to remove the two merge commits before completing the PR as a rebase & ff PR here.

Comment (apop5) [reply UefiCpuPkg/CpuMpPei/CpuPaging.c:1]: @makubacki 

Fixed the cherry-pick to remove the merges, if you could review it.

**Review by makubacki — APPROVED** [review]

**Pull request merged**